File: src/utils/AnthropicProvider.js
/**
 * AnthropicProvider - Claude API provider for intelligence features
 * 
 * Note: Anthropic doesn't provide embeddings (yet), so this provider
 * is for intelligence features only (auto-tag, synthesis, etc).
 * 
 * Pair with another provider for embeddings:
 *   - Use OpenAI/Ollama/Gemini for embeddings
 *   - Use Anthropic for intelligence
 * 
 * Environment:
 *   ANTHROPIC_API_KEY=your_api_key
 *   ANTHROPIC_MODEL=claude-3-haiku-20240307
 * 
 * Get API key: https://console.anthropic.com
 */

import { IntelligentProvider } from '../interfaces/EmbeddingProvider.js';

export class AnthropicProvider extends IntelligentProvider {
  constructor(options = {}) {
    super({
      name: 'Anthropic',
      dimensions: options.dimensions || 1536,
      ...options
    });
    
    this.apiKey = options.apiKey || process.env.ANTHROPIC_API_KEY;
    this.model = options.chatModel || process.env.ANTHROPIC_MODEL || 'claude-3-haiku-20240307';
    this.baseUrl = options.baseUrl || 'https://api.anthropic.com/v1';
    
    // For embeddings, we need a fallback provider
    this.embeddingFallback = options.embeddingFallback;
  }

  isAvailable() {
    return !!this.apiKey;
  }

  /**
   * Anthropic doesn't provide embeddings API yet.
   * This will use the fallback provider or return null.
   */
  async embed(text) {
    if (this.embeddingFallback) {
      return this.embeddingFallback.embed(text);
    }
    
    console.error(
      '[AnthropicProvider] Embeddings not supported. ' +
      'Use a different provider for embeddings or set embeddingFallback.'
    );
    return null;
  }

  async embedBatch(texts) {
    if (this.embeddingFallback) {
      return this.embeddingFallback.embedBatch(texts);
    }
    return texts.map(() => null);
  }

  async autoTag(key, content) {
    if (!this.apiKey) return [];

    const prompt = `Generate 5-10 relevant tags for this content.
Return ONLY a JSON array of lowercase, hyphenated tags like ["tag-one", "tag-two"].

Title: ${key}
Content: ${content.substring(0, 3000)}`;

    try {
      const result = await this.createMessage(prompt);
      const cleaned = result
        .replace(/```json/g, '')
        .replace(/```/g, '')
        .trim();
      return JSON.parse(cleaned);
    } catch (err) {
      console.error(`[AnthropicProvider] Auto-tag failed: ${err.message}`);
      return [];
    }
  }

  async synthesize(content, task, style = '') {
    if (!this.apiKey) {
      return content.substring(0, 1500);
    }

    const styleGuides = {
      beginner: 'Explain simply for beginners using analogies.',
      detailed: 'Be comprehensive with examples and edge cases.',
      concise: 'Be extremely brief. One paragraph maximum.'
    };

    const system = `You are a helpful assistant. ${styleGuides[style] || ''}`;
    const prompt = `Context from notes:\n${content.substring(0, 8000)}\n\nTask: ${task}`;

    try {
      return await this.createMessage(prompt, system);
    } catch (err) {
      console.error(`[AnthropicProvider] Synthesis failed: ${err.message}`);
      return content.substring(0, 1500);
    }
  }

  async extractGraphEntities(content) {
    if (!this.apiKey) return { concepts: [], relationships: [] };

    const system = 'You extract knowledge graph data. Return ONLY valid JSON.';
    const prompt = `Extract key concepts and relationships from this text.
Return JSON format:
{
  "concepts": ["concept1", "concept2", ...],
  "relationships": [
    {"from": "concept1", "to": "concept2", "type": "related_to", "strength": 0.8}
  ]
}

Text: ${content.substring(0, 4000)}`;

    try {
      const result = await this.createMessage(prompt, system);
      const cleaned = result
        .replace(/```json/g, '')
        .replace(/```/g, '')
        .trim();
      const parsed = JSON.parse(cleaned);
      return {
        concepts: Array.isArray(parsed.concepts) ? parsed.concepts : [],
        relationships: Array.isArray(parsed.relationships) ? parsed.relationships : []
      };
    } catch (err) {
      console.error(`[AnthropicProvider] Graph extraction failed: ${err.message}`);
      return { concepts: [], relationships: [] };
    }
  }

  async createMessage(prompt, system = '') {
    const body = {
      model: this.model,
      max_tokens: 2048,
      messages: [{ role: 'user', content: prompt }]
    };

    if (system) {
      body.system = system;
    }

    const response = await fetch(`${this.baseUrl}/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': this.apiKey,
        'anthropic-version': '2023-06-01'
      },
      body: JSON.stringify(body)
    });

    if (!response.ok) {
      throw new Error(await response.text());
    }

    const data = await response.json();
    return data.content?.[0]?.text || '';
  }

  getConfig() {
    return {
      type: 'anthropic',
      model: this.model,
      dimensions: this.dimensions,
      hasApiKey: !!this.apiKey,
      hasEmbeddingFallback: !!this.embeddingFallback,
      note: 'Requires embeddingFallback for embeddings'
    };
  }
}

File: src/adaptive_memory_engine/providers/anthropic_provider.py
"""Anthropic provider (chat only — embeddings require a fallback)."""

from __future__ import annotations

import json
import logging
import re
from collections.abc import Iterable

import httpx

from .base import EmbeddingProvider, IntelligentProvider
from .http import request_with_retry

log = logging.getLogger(__name__)


class AnthropicProvider(IntelligentProvider):
    name = "anthropic"
    dimensions = 1536  # virtual; we delegate embedding to fallback

    def __init__(
        self,
        api_key: str | None,
        embedding_fallback: EmbeddingProvider,
        base_url: str = "https://api.anthropic.com/v1",
        chat_model: str = "claude-haiku-4-5-20251001",
        embedding_model: str = "claude-haiku-4-5-20251001",
        timeout: float = 60.0,
    ) -> None:
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.chat_model = chat_model
        self.embedding_model = embedding_model
        self._fallback = embedding_fallback
        self._client = httpx.Client(timeout=timeout)
        # Use the fallback's true dimensions
        if hasattr(embedding_fallback, "dimensions"):
            self.dimensions = embedding_fallback.dimensions

    def is_available(self) -> bool:
        return bool(self.api_key) and self._fallback.is_available()

    def embed(self, text: str) -> list[float]:
        return self._fallback.embed(text)

    def embed_batch(self, texts: Iterable[str]) -> list[list[float]]:
        return self._fallback.embed_batch(texts)

    def auto_tag(self, key: str, content: str) -> list[str]:
        text = self._messages(
            "Generate 3-5 concise lowercase tags as a JSON array of strings. "
            "Reply with JSON only.\n\n"
            f"Key: {key}\nContent: {content[:1500]}\nTags:"
        )
        return self._parse_tag_list(text)

    def synthesize(self, content: str, task: str, style: str = "concise") -> str:
        return self._messages(
            f"Task: {task}\nStyle: {style}\n\nContent:\n{content[:6000]}\n\nResponse:"
        )

    def expand_query(self, query: str) -> list[str]:
        text = self._messages(
            f"Generate 3 paraphrases as a JSON array of strings.\n\nQuery: {query}\nParaphrases:"
        )
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("queries") or arr.get("paraphrases") or []
            if isinstance(arr, list):
                return [str(x) for x in arr]
        except Exception:  # noqa: BLE001
            pass
        return [query]

    def get_config(self) -> dict:
        return {
            "type": "anthropic",
            "chatModel": self.chat_model,
            "embeddingFallback": getattr(self._fallback, "name", "unknown"),
        }

    def _messages(self, prompt: str) -> str:
        r = request_with_retry(
            self._client,
            "POST",
            f"{self.base_url}/messages",
            headers={
                "x-api-key": self.api_key or "",
                "anthropic-version": "2023-06-01",
                "Content-Type": "application/json",
            },
            json={
                "model": self.chat_model,
                "max_tokens": 1024,
                "messages": [{"role": "user", "content": prompt}],
            },
        )
        r.raise_for_status()
        content = r.json()["content"]
        return "".join(b.get("text", "") for b in content if b.get("type") == "text")

    def close(self) -> None:
        self._client.close()

    def _parse_tag_list(self, text: str) -> list[str]:
        try:
            arr = json.loads(text)
            if isinstance(arr, dict):
                arr = arr.get("tags", [])
            if isinstance(arr, list):
                return [str(t).strip().lower() for t in arr if str(t).strip()][:10]
        except Exception:  # noqa: BLE001
            pass
        return [m.group(1).lower() for m in re.finditer(r'"([^"]+)"', text)][:10]
